Firestore/core/include/firebase/firestore/timestamp.h

```diff
@@ -216,6 +217,27 @@ struct hash<firebase::Timestamp> {
   // implementation is subject to change.
   size_t operator()(const firebase::Timestamp& timestamp) const;
 };
+
+template <>
+class numeric_limits<firebase::Timestamp> {
```

Comment:
So this is ... curious. :) Specializing std::numeric_limits with our types works and is an interesting way to surface this. But it's not clear to me if this is either (a) a good idea, or (b) even legal. (The spec's a little vague on this for non-standard types.) We could trivially move this to `static Timestamp::Max()` or similar. If we choose to keep it here, we'd eventually want to fill it out a bit more.

Having this available also implies some of the Timestamp implementation and tests could take advantage of this too, but I haven't adjusted that in this PR. (It could be a followup.)

Another alternative might be to just redefine these values (again) in the serializer (but then we'd have 3 copies of it floating around.)

Yet another alternative would be to add a factory method that returns a StatusOr (which returns a bad status rather than asserting when values are too large) and use that rather than the ctor in the serializer.

Options tl;dr:
1) specialize std::numeric_limits
2) Timestamp::Max()/Min()
3) Copy kUpperBound to serializer.cc (and don't touch Timestamp)
4) static StatusOr<Timestamp> Timestamp::Create(seconds, nanos)

Reply:
I like having `max` and `min`. As for selecting between 1 and 2... On my reading, it also appears that specializing `numeric_limits` is a little murky. It appears to [be allowed](https://stackoverflow.com/questions/15305723/difference-between-stdis-signedt-and-stdnumeric-limitstis-signed#comment21604593_15305830). OTOH, it might be more trouble than it's worth:
* referring to the standard for guidance, `<chrono>`'s `time_point` which is reasonably similar to `Timestamp` doesn't specialize `numeric_limits`. This is of course non-binding.
* I don't think that requirements from the standard apply to specialization for a user-defined type, but not implementing them will make the specialization inconsistent with the standard, and implementing them is a _lot_ of boilerplate (defining other member functions, providing const/non-const specializations).
* just in general, it appears that specializing `numeric_limits` for user-defined types is underspecified and probably not expected to be a common occurrence (unlike e.g. `hash`).

So all in all, I'd vote for option 2. For its pros and cons:
* since the notion of maximum and minimum time is pretty explicit in `Timestamp`, I think it's reasonable to encode that;
* OTOH, I'm not sure if this change would require a round of review (the only way around that would be option 3, though).

For option 4, I'm wary because this will entail making Firebase common depend on `StatusOr`, which might be undesirable. We could, however, define this function as a helper just within Firestore.

Reply:
Just to brainstorm:
* there could also be a function (member or free) like `IsWithinSupportedRange` (although it would make it harder to report the error precisely).
* these could be public constants (e.g., `kMaxSeconds`).

Reply:
> For option 4, I'm wary because this will entail making Firebase common depend on StatusOr, which might be undesirable. We could, however, define this function as a helper just within Firestore.

Good point. We could migrate firebase::firestore::Status to firebase::Status... but I'm not sure I want to take that on.

> OTOH, I'm not sure if this change would require a round of review (the only way around that would be option 3, though).

Yeah, I was a little concerned about that too. However, this issue *might* only exist in c++? For instance, in java, anyone who cares about the range can just try it and "ask for forgiveness rather than permission". IOW, they can catch the resulting exception. (And since *all* of the c++ sdk needs to eventually get reviewed...) But I'm pointing this out from more of a theoretical viewpoint; if we're going to expose a max/min, then I'd like to do so across platforms.

Anyways, since the rest of the issues so far are nits, I'm going to drag @wilhuff in now to get his opinion too.

(I'm leaning toward #2 too.)

Reply:
FWIW, my general take is that unless something is described as the way to do something (and there's lots of prior art) it's best to stay away from extending the standard library. The essential problem is that it puts us in a position where things can fail for obscure reasons on compilers/libraries we don't use often for very little benefit.

If someone really, really needed a specialization of `numeric_limits` it's trivial to implement given `Min` and `Max` functions we supply, but if our specialization of `numeric_limits` interacts poorly with something else in a user system it's very difficult to work around.

Secondarily, and more specific to this case: `std::numeric_limits` is a poor fit for what we're doing here. Those traits include members like `epsilon`, `infinity`, and `quiet_NaN` (http://en.cppreference.com/w/cpp/types/numeric_limits). `std::numeric_limits` defines reasonable behavior for all of the specializations in the standard library and we can't reasonably do so for `Timestamp`. This makes it unlikely that this specialization will work for any algorithm that might consume a type based on `std::numeric_limits` traits.

I'm also not a fan of (2) because it's a committed API for which we don't necessarily know there's a demonstrable need.

My proposal is a variation on (3) (call it (5)):
Put internal helpers for a type that aren't part of the committed API in a separate header, e.g. right next to `timestamp.cc` add `timestamp_internal.h` that includes whatever you need to share between `timestamp.cc` and the serializer. Be mindful of how this is shared on the other platforms (if at all).